\documentclass{article}

% -- Packages
\usepackage{blindtext}
\usepackage[utf8]{inputenc}
\usepackage{biblatex}
\usepackage{hyperref}
\usepackage{parskip}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{microtype}
\usepackage{pdfpages}
\usepackage{{booktabs}}
\usepackage{{multirow}}
\usepackage{geometry}
\usepackage{bookmark}
\usepackage{multicol}
\usepackage{mathtools}
\usepackage{float}

% -- Theorem environments
\newtheorem{definition}{Definition}
\newtheorem{example}{Example}
\newtheorem{remark}{Remark}

% -- Document Settings
\emergencystretch=1em
\title{Deriving BLOSUM-like matrices for protein-coding DNA}
\geometry{a4paper, left=15mm, right=15mm, top=15mm, bottom=25mm}
\addbibresource{substitution-matrices.bib}
\addbibresource{software.bib}
\bibliography{substitution-matrices}
\twocolumn
\author{Alexander Temper}

\begin{document}
\maketitle

\begin{abstract}
    Sequence Alignment algorithms rely on some form of scoring matrix,
    which can heavily influence their results. In this work, we try to derive
    scoring matrices for the deoxyribonucleic acid (DNA) of subsets of protein
    families for specific organisms akin to
    the BLOSUMx matrices, which are derived for amino acid sequences.

    Whilst the lack of proper benchmark data allows for no definitve quantitative
    statement about the performance of the derived matrices, we find that there is ample diversity among them to be further investigated.
\end{abstract}

\tableofcontents

\section{Introduction} Sequence alignment is the task of pairing up the
elements of (biological) sequences with the aim of finding conserved
regions and homologous sequences anticipated to have some
evolutionary relation. Sequence alignment algorithms are primarily used for
constructing phylogenetic trees \cite{chatzouMultipleSequenceAlignment2016}, searching for homologous sequences with BLAST \cite{altschulBasicLocalAlignment1990},
assembling full DNA strands from short reads, creating input features for
machine learning algorithms like AlphaFold2 \cite{jumperHighlyAccurateProtein2021a} and others \cite{chatzouMultipleSequenceAlignment2016}

Most alignment algorithms try to maximize a \emph{scoring function},
which assigns a measure of supposed quality to a given alignment.
The classical methods score
alignments by summing up scores assigned to each pair of letters in the given
alignment. The scores for the pairs of letters can be represented by a
symmetric matrix --- we call such matrices "\emph{scoring matrices}".

\begin{definition}
    The following will be used throughout the paper:
    \begin{itemize}
        \item A \emph{sequence} $s$ is an ordered collection of letters from
              an alphabet $\mathcal{L}$.
        \item The letter '\texttt{-}' is called \emph{indel} and represents
              gaps in the alignment, which typically occur due to mutations or sequencing
              errors.

        \item A (sequence) alignment is a matrix $\mathbf A \in
                  (\mathcal{L} + \{\mathtt{-}\})^{n \times m}$, where $n$ is the number of aligned sequences and $m$ is the
              length of the alignment.
	      An alignment with $n = 2$ is called a \emph{pairwise sequence alignment.}
      \item A \emph{scoring function} $\sigma: (\mathcal{L} + \{\mathtt{-}\})^{n
                      \times m} \to \mathbb{R}$ maps an alignment to its score.
      \item A pairing of two nuleotides $a, b$  in an alignment is called a "\emph{match}" if $a = b$, a "\emph{mismatch"} if $a \ne b$ and a "\emph {gap}" if $a =$ "\texttt{-}" or $b =$ "\texttt{-}".
    \end{itemize}
\end{definition}


The classical algorithms for pairwise sequence alignment use a scoring function of the form
\begin{equation*}
	\sigma_s( \mathbf A )= \sum_{j=0}^m s(\mathbf A_{1j}, \mathbf A_{2j}),
\end{equation*} where $s:
    (\mathcal L + \{\mathtt{-}\})^2 \to \mathbb{R}$ is a symmetric function
evaluating pairs of letters, which can be represented by a matrix $\mathbf S
    \in \mathbb Z^{(\# \mathcal L) + 1 \times (\# \mathcal L) + 1}$. Said matrix is
the focus of this work.

\begin{example}
    Our alphabet of concern will be the 4
    different nucleotide bases, \{\texttt{A, C, G, T}\}. Two exemplary DNA
    sequences are \texttt{ACA} and \texttt{AAGA}. One possible alignment between them is
    \begin{equation*}
        \mathbf A = \begin{bmatrix}
            \mathtt{A} & \mathtt{C} & \mathtt{-} & \mathtt{A} \\
            \mathtt{A} & \mathtt{A} & \mathtt{G} & \mathtt{A}
        \end{bmatrix}.
    \end{equation*}
    An exemplary scoring matrix is 
    \begin{equation*}
        \mathbf S = \begin{matrix}
             & \mathtt{A} & \mathtt{C} & \mathtt{G} & \mathtt{T} &
            \mathtt{-}                                             \\ \mathtt{A} & 1          & -2         & -3         & 0
             &
            0
            \\ \mathtt{C} & -2         & 0          & 0          & 0
             & 0
            \\ \mathtt{G} & -3         & 0          & 0          & 0
             /& -4
            \\ \mathtt{T} & 0          & 0          & 0          & 0
             & 0
            \\ \mathtt{-} & 0          & 0          & -4         & 0
             & 0 \\
        \end{matrix}.
    \end{equation*}
    Using the above matrix, \begin{align}
	    \sigma_{\mathbf S}(\mathbf A) &= s(\mathtt A, \mathtt A) + s(\mathtt C, \mathtt A) + s(\mathtt{-}, \mathtt{G}) + s(\mathtt{A}, \mathtt{A}) \\
					  &= 1 - 2 -4 + 1 \\
					  &= -6.
    \end{align}
\end{example}

\subsection{BLOSUM}
Famously, one family of scoring matrices are called BLOSUMs (\textbf{BLO}ck \textbf{SU}bstitution \textbf{M}atrices), first constructed by \textcite{henikoffAminoAcidSubstitution1992} for aligning amino acid sequences. A wonderful explanation thereof was written by
\textcite{eddyWhereDidBLOSUM622004}. However, we shall briefly dive into the
theoretical underpinnings of BLOSUMs here as well.

Underlying BLOSUMs is the equation
\begin{equation*}
	s(a, b)= 2 \log_2 \frac{P(a, b)}{P(a)P(b)} \text{ where } a  \in \mathcal L, b \in \mathcal L.
\end{equation*}

We can understand this as follows: We want to derive a score we assign
for aligning the two letters $a$ and $b$, $s(a,b)$. We can represent such
a score by the odds of the probability of aligning $a, b$ being observed,
$P(a, b)$, versus the probability of the two being paired due to chance. In probablistic terms this is $P(a \cup b) = P(a)P(b)$.

Now, assuming we have a dataset with already aligned data we deem to be good, we
can approximate $P(a, b)$ by counting how often $a$ and $b$ were aligned, and
we can also approximate $P(a)P(b)$ by counting how often both $a$ and $b$ appear
individually and multiplying the resulting two counts together.

In this fashion, the orgiginal BLOSUM was constructed
from the BLOCKS database \cite{henikoffAutomatedAssemblyProtein1991}, which provided gapless multiple sequence alignments (MSAs) called BLOCKS, frrom which $P(a,b)$ and $P(a)$ and $P(b)$ were approximated in a frequentist fashion. Given the absence of gaps, the re-
sulting scoring matrix does not include scores for gaps. We
now follow up with a somewhat precise definition of a simplified version of the algorithm to compute the BLOSUM.

\begin{definition}
A BLOCKS database $\mathbf Z$ is a sequence of MSAs of differing shapes, i.e., $\mathbf Z = (\mathbf A^k)_{k=1}^l$
	Let $\mathbf C \in \mathbb N^{4 \times 4}$ be a symmetric matrix
with counts of how often each nucleotide was aligned to another nucleotide in each column of each BLOCK.
Let $\mathbf Q$ be the matrix of the relative frequencies of the observed pairs, i.e.,
\begin{equation*}
	 \mathbf Q_{i, j} = \mathbf C_{i, j} / \sum_i^4 \sum_j^i \mathbf C_{i, j}j
\end{equation*}
Further, we denote the relative frequencies of each nucleotide in $\mathbf Z$ as $\mathbf p$, and it can
be computed from $\mathbf Q$ via $\mathbf p_i = \mathbf Q_{i, i} + \sum_{j \ne i, j = 0}^4 \frac {\mathbf Q_{i, j}} {2}$.
\end{definition}
\begin{example}
	Let $\mathbf Z$ consist of two BLOCKS:
	\begin{equation*}
		\begin{matrix}
		\begin{matrix}
			\texttt{AA} \\
			\texttt{AC} \\
			\texttt{AC} \\
		\end{matrix} &
		\begin{matrix}
			\texttt{AA} \\
			\texttt{AC} \\
			\texttt{AC} \\
		\end{matrix}.
	\end{matrix}
	\end{equation*}
	For the first block in the first column, we have three pairs of \texttt{AA}, and in the second column we have two pairs of \texttt{AC} and one pair of \texttt{CC}
	The second  is identical, so the counts are identical. Assuming the indexes $\mathtt A = 1, \mathtt C = 2, \mathtt G = 3, \mathtt T = 4$, we get
	\begin{equation*}
		\mathbf C = \begin{bmatrix}
			6 & 4 & 0 & 0 \\
			4 & 2 & 0 & 0 \\
			0 & 0 & 0 & 0 \\
			0 & 0 & 0 & 0
		\end{bmatrix}
	\end{equation*}
	thus
	\begin{equation*}
		\mathbf Q = \begin{bmatrix}
			0.5 & 0.33 & 0 & 0 \\
			0.33 & 0.16 & 0 & 0 \\
			0 & 0 & 0 & 0 \\
			0 & 0 & 0 & 0
		\end{bmatrix}
	\end{equation*}
	and
	\begin{equation*}
		\mathbf p = \begin{bmatrix}
			0.66 & 0.33 & 0 & 0
		\end{bmatrix}.
	\end{equation*}
\end{example}

Now the assumption is that, given that $i$ is the index of $a$ and $j$ is the index of $b$,
\begin{equation*}
	s(a, b) \approx	2 \log_2 \begin{cases}
		\frac{\mathbf Q_{i, j}}{2p_ip_j} & i \ne j, \\
		\frac{\mathbf Q_{i, j}}{p_i^2} & i = j
	\end{cases}
\end{equation*}
--- rounding these values to the nearest integer then gives us our scoring matrix $\mathbf S$.

\begin{remark}
	One might rightfully ask why we apply a binary logarithm and a doubling to the odds.
	The logarithm of odds, called \emph{log-odds}, are fairly standard in
	statistics --- odds are by definition positive, and opposing odds, i.e.,
	those where the $\frac a b < 1$, span only
	$(0, 1)$, whereas good odds have a range of $[1, \infty)$. Applying a logarithm counteracts
	this by making the function skew symmetric, i.e., $\frac a b 
	\ne -\frac b a$ but $\log \frac a b = - \log \frac b a $.
	
	To the best of our knowledge, there is no distinct reason why the number 2 is picked for the logarithm and the multiplier - we suspect this is due to its connection to information theory.
\end{remark}

An important piece of the algorithm is still missing - the clustering of sequences within BLOCKS.
We will introduce this more informally: we pick a similarity threshold $x$, we
then check within each  which sequences have $\ge x$\% identity and
then ``average'' the sequences transitively, i.e., if sequence $i$ and
sequence $j$ are to be clustered and sequence $j$ and sequence $k$ as well,
then sequence $i$, sequence $j$ and sequence $k$ are being clustered together.
The count matrix $\mathbf C$ now includes fractional counts, and the rest of the computation remains the same.
This is why the matrices are called BLOSUMx matrices --- a BLOSUM90 matrix is made from blocks where sequences 
of 90\% similarity are clustered.

\begin{example}
	Let
	\begin{equation*}
		\mathbf A = \begin{matrix}
			\mathtt{AAC} \\
			\mathtt{AGC} \\
			\mathtt{AGG} \\
			\mathtt{TTT}
		\end{matrix}	
	\end{equation*}
	and $x = \frac 2 3$. Sequences $(1, 2)$ and $(2, 3)$ exceed the similarity threshold,
	thus the clustered  will be
	\begin{equation*}
		c(\mathbf A) = \begin{matrix}
		\mathtt{A}(\frac 2 3 \mathtt G \frac 1 3 \mathtt A)(\frac 2 3 \mathtt C \frac 1 3 \mathtt G) \\	
		\mathtt{TTT}
		\end{matrix} 	
	\end{equation*}
	The resulting count matrix 
	\begin{equation*}
			\mathbf C = \begin{bmatrix}
			0 & 0 & 0 & \frac 4 3 \\
			0 & 0 & 0 & \frac 2 3 \\
			0 & 0 & 0 & 1 \\
			\frac 4 3 & \frac 2 3 & 1 & 0
		\end{bmatrix}.
	\end{equation*}
\end{example}
This amounts to the basic machinery of the algorithm to compute BLOSUMs.

\begin{remark}
	This family of matrices has become a de-facto standard for aligning amino acid
sequences - however, not so much for DNA. BLASTn, i.e., BLAST for nucleotide bases, currently uses a matrix where
matches are rewarded with +2 and mismatches are penalized with -3 \cite{altschulBasicLocalAlignment1990}
very assumed generality that has motivated this paper.
\end{remark}

\begin{remark}
Interestingly enough, there have been
mistakes in the original computation of the BLOSUMs, which are claimed
to have been improving them quietly
\cite{styczynskiBLOSUM62MiscalculationsImprove2008a}. However, this might be
attributed to the fact that the benchmarks today might be influenced from the BLOSUMs of the past.
\end{remark}

\begin{remark}\label{emref}
	The reader might notice that there is some form of circular reasoning going on around
	the computation of the BLOSUMs - good alignments are produced by using BLOSUMs 
	and BLOSUMs are produced by good alignments. Indeed, the BLOCKS database
	itself is created with BLOSUMs, and the two iteratively create the other.

	First, the BLOCKS database is created from a scoring matrix with 0
	mismatch score and 1 match score. Then, the first BLOSUM is created,
	which is then used to produce new BLOCKS. This process is then repeated 
 	an additional two times until the actual BLOSUM and BLOCKS database have been
	produced.
\end{remark}

\subsection{Related work}
Most notably, the two families of classical scoring matrices are the PAM matrixes \cite{dayhoffModelEvolutionaryChange1978} and the BLOSUMs. 
However, and this is also true for most other research around sequence alignment: there is a strong emphasis on studying the
alignment of proteins over studying the alignment of DNA. To the best of our
knowledge, nearly all benchmarks concern amino acid alignments \cite{gardnerBenchmarkMultipleSequence2005, pervezEvaluatingAccuracyEfficiency2014, thompsonBAliBASELatestDevelopments2005, vanwalleSABmarkBenchmarkSequence2005a}, which
we suspect to to be one of the reasons that there is limited research focusing
on DNA alignment. There have been attempts to create gold standards for RNA
alignment \cite{wilmEnhancedRNAAlignment2006},
however, even there, investigations suggest an overrepresentation of tRNA, thus
leading to a suboptimal benchmark \cite{lowesBRaliBaseDentTale2017}

The literature for DNA scoring matrices derived from data is sparse.
\textcite{hamadaTrainingAlignmentParameters2017} derive scoring matrices from
specific organisms sequenced by specific sequencers. Further, as they claim,
different genes in different organisms have significantly differing rates of
mutation, which is why general-purpose DNA scoring matrices might be a bad
idea. Their main focus lies on recovering and mitigating the sequencing errors
and projecting the differing GC contents of different species in the resulting
matrix. They evaluate their data on simulated data, which consequentually makes
strong assumptions. Thus, their findings are not  too applicable to real world data.
Moreover, their method performs only slightly better than 2 manually created scoring matrices, which we deem to be insufficient evidence to support any substantial claims.


\section{Experiments and results}
\subsection{Method}
We constructed a fully automated pipeline which takes as input an identification code of a protein
family on InterPro and the similarity threshold $x$ and computes the
corresponding nBLOSUMx (nucleotide BLOSUMx). The source code can be found \href{https://github.com}{online}, yet, here we
give some details on the implementation and issues we faced.

First, we search the \href{https://www.ncbi.nlm.nih.gov/protein}{NCBI protein
database} for the given identification code and given organism using eDirect
using the query \texttt{<ORGANSIM> [ORGN] AND <INTERPRO\_CODE>}.
The following protein families were selected from InterPro:
\input{plots/protein_families.tex} 
This yields only a subset of the desired genes, since not every
protein in the database is annotated with all protein family codes and its organism. This method is,
however, to the best of our knowledge, the fastest and most reliable way to
download genes of a given protein family on InterPro --- for the scope of this
work, we deem this sufficient.

Next, the downloaded genes are preprocessed. Genes which contain letters 
besides \texttt{A, C, G, T} are removed. Afterwards, they are being sorted into bins,
depending on their length, the rationale thereof being that most multiple
sequence alignmeners assume that the sequences are of similar length. Further,
bins consisting of too few sequences are being removed.

Unfortunately, the PROTOMAT \cite{henikoffAutomatedAssemblyProtein1991} found online any longer. 
Thus, we create our own BLOCKS, which is definitely a major drawback and flaw of this work. 

To do so, each bin is getting aligned by kalign \cite{lassmannKalignAccurateFast2005} with
its default settings. There was no specific rationale in picking kalign - it is
user friendly, fast and reasonably accurate. Aligning the bins results
in a multiple sequence alignment for each bin.

Thereafter, each MSA is looked at and gapless regions are selected. 
These will be the BLOCKS for the computation of the BLOSUMx.
Subsequently, we compute the BLOSUM90s of the created BLOCKS - note that $x$
was chosen somewhat arbitrarily, with the only reason that lower values would
cluster too many sequences for narrower organism groups. Also note
that we do not iteratively refine the BLOCKS as noted in Remark \ref{emref}
as we were unable to find a multiple sequence aligner that allows an easy specification
of the scoring matrix.
\subsection{Results \& Discussion}	
	Multiple nBLOSUM90s have been derived, and the main takeaway is that the they differ substantially from the +2/-3 matrix employed by BLASTn. 	

	In the beginning, no filtering by organism has been undertaken, leading 
	to what was essentially the identity matrix, i.e., rewarding matches
	with +1 and not penalizing mismatches. We suspect this behavior due to
	the vast diversity of organisms within each protein family - said
	diversity implies large evolutionary distance between sequences of
	BLOCKS, leading to what is essentially noise within the multiple
	sequence alignments. It would be nearly impossible to align full
	protein families.
	
	Constructing the BLOSUMs for specific taxa, however, lead to
	more refined scoring matrices which show some interesting patterns.

	First and foremost, the different mismatch scores were rarely uniform, indicating that a general mismatch score might be suboptimal for aligning sequences of known taxa. 	Further, we noticed that match scores were typically similar, but not
	equal. This might be attributable to noise in the data, as
	often the diagonals were indeed repeating the same score. 

	Another observation to be made is that, transitions,
	i.e., mismatches of \texttt{A} $\iff$ \texttt{G} and \texttt{T}
	$\iff$ \texttt{C}, ended up receiving a smaller penalty than
	transversions, i.e., all other mismatches. There are models of
	evolution making this assumption \cite{kimuraEstimationEvolutionaryDistances1981}, which is called
	the transition bias phenomenon, and
	finding this within derived scoring matrices is a supporting argument thereof.

	Further, another observation is that the more diverse the examined taxa,
	the lower the resulting scores. As can be seen below,
	the matrices of animals and bacteria have the lowest standard deviation and norm (indicating low values in the matrix) --- this does, however, not hold for plants, which are also largely diverse. 
	
As a general theme, we would like to stress that the matter of large scale
genetic data is very complex, and given the lack of a gold standard, our still
reasonably small understanding of genetics and the scope of this work no
substantial statements can or should be made.
	\begin{table}[h] 
		\include{plots/table.tex}
		\caption{Descriptive statistics about the derived nBLOSUM90s.
			``\# Seq.'' is the number of sequences used for
			constructing each matrix, $\sigma^2(S)$
is the standard deviation of all 16 scores per matrix and $\| \mathbf S\|_2$ is the Frobenius norm.}
	\end{table}

	\begin{table}
	\input{plots/matrices.tex}
	\caption{The derived nBLOSUM90 matrices, per organism and protein family}
	\end{table}
\onecolumn
\pagebreak

\begin{figure}[!htp]
		\centering
		\includegraphics[width=\textwidth]{plots/downprojection.pdf}
		\caption{Various nBLOSUMs derived from the specific protein
		families for different subsets of species, projected to two
		dimensions using Principal Component Analysis
	\cite{leverPrincipalComponentAnalysis2017} plotted with plotly
\cite{plotly} and scikit-learn \cite{scikit-learn}. The species is indicated
by color and protein family by symbol.}
	\end{figure}

Visually inspecting the downprojection of the BLOSUMs , we can also find
hints for the fact that it is rather the species that should influence scoring
matrices, rather than specific protein families. Note that the derived matrices
for different protein families for a given taxa lie close together in the plot,
whereas the matrices of some protein families spread rather far apart. This can also be
seen inspecting the tables themselves.

\section{Conclusion}
We conclude with a few insights. First and foremost, that the most 
needed advance in the field may be a reproducible and generally agreed upon
consensus of what makes a good alignment for DNA, and a resulting benchmark
dataset.

Further, whilst no quantitative superiority of derived scoring matrices for DNA
has been established, observed differences between the currently employed matrix by BLASTn and the derived matrices (which do have some underlying theory as
opposed to the empirically chosen match-/mismatch-scores used by BLASTn) give
reason to further investigate data driven scoring matrices. Moreover, the
online user interface of BLAST should allow for user defined scoring matrices
for, first to raise awarness of the arbitrariness of the current options and
secondly to allow for further development of new nucleotide substitution
matrices.

\pagebreak
\nocite{reback2020pandas}
\printbibliography
\end{document}
